woocommerce/woocommerce-ios
# Fix an issue where the same gift card ID cannot be shown in multiple orders

Prerequisite for #10518 


## Why


While working on gift card support during order creation/editing, I noticed that the updated order's gift cards disappear instantly even though the `gift_cards` field is non-empty in the API response. I was quite stuck as everything seemed to be fine, then when I inspected the database there were only 3 gift card records (my site only has 3 unique gift cards). It turned out in `OrdersUpsertUseCase.handleOrderGiftCards`, it assumes that there can be only one `OrderGiftCard` record of each ID across the site. This means that if there are multiple orders with the same gift card applied, the gift card can only be shown in one of the orders. This PR fixes the behavior by allowing any number of gift cards of the same ID for each order. The merchant can edit the order and apply the same gift card that has a higher balance.

## How

In `OrdersUpsertUseCase.handleOrderGiftCards`, the existing storage gift cards are always deleted first and then readonly gift cards are inserted to the storage.

## Testing instructions


Prerequisite: the site has the [Gift Cards](https://woocommerce.com/products/gift-cards/) extension/plugin, and at least one gift card code available with some remaining amount.

- On the shop web page, place two orders using the same gift card code
- Launch the app
- Go to the Orders tab
- Tap on the first order with the gift card --> the gift card code should be shown with the correct amount
- Go back to the Orders tab
- Tap on the second order with the gift card --> the gift card code should also be shown with the correct amount (in `trunk`, only one order shows the gfit card)

## Screenshots


order 1 | order 2
-- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/f0ec65fe-ec23-4754-b312-7e0ced394c2b" width="300" />  | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/2557a899-e356-4735-abba-180d96180e9d" width="300" />

---
